stratosphere/loader/source/ldr_capabilities.cpp

```diff
@@ -0,0 +1,428 @@
+                        const u32 restrict_size = restrict_next.GetAddressSize();
+                        const u32 restrict_end = restrict_start + restrict_size;
+
+                        if (restrict_size >= SizeMax) {
+                            continue;
+                        }
+
+                        if (this->GetFlag() == restrict.GetFlag() && next.GetFlag() == restrict_next.GetFlag()) {
+                            if (restrict_start <= start && start <= restrict_end && end <= restrict_end) {
+                                return true;
+                            }
+                        }
+                    }
+                }
+                return false;
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(MapPage,
+            u32 GetAddress() const {
+                return (this->GetValue() >> 0) & 0xFFFFFF;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        if (this->GetValue() == restrict.GetValue()) {
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(InterruptPair,
+            Result Validate(const u32 *kac, size_t kac_count) const;
+
+            static constexpr u32 EmptyInterruptId = 0x3FF;
+
+            u32 GetInterruptId0() const {
+                return (this->GetValue() >> 0) & 0x3FF;
+            }
+
+            u32 GetInterruptId1() const {
+                return (this->GetValue() >> 10) & 0x3FF;
+            }
+
+            bool IsSingleIdValid(const u32 id, const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        if (restrict.GetInterruptId0() == EmptyInterruptId && restrict.GetInterruptId1() == EmptyInterruptId) {
+                            return true;
+                        }
+
+                        if (restrict.GetInterruptId0() == id || restrict.GetInterruptId1() == id) {
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                return IsSingleIdValid(this->GetInterruptId0(), kac, kac_count) && IsSingleIdValid(this->GetInterruptId1(), kac, kac_count);
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(ApplicationType,
+            u32 GetApplicationType() const {
+                return (this->GetValue() >> 0) & 0x3;
```

Comment:
This is pretty much a personal choice, but I prefer not to use `>> 0` since it doesn't do anything. The bitmasking is sufficient for me to understand that this is extracting bits starting at bit0.

Reply:
Yeah, I prefer >> 0 for aesthetic symmetry reasons.